components/browser/storage-sync/src/main/java/mozilla/components/browser/storage/sync/Types.kt

```diff
@@ -66,3 +70,51 @@ internal fun mozilla.appservices.places.VisitInfo.into(): VisitInfo {
         visitType = this.visitType.into()
     )
 }
+
+/**
+ * Holds fields common to all Glean sync engine pings.
+ */
+internal data class BaseGleanSyncPing(
+    val uid: String,
+    val startedAt: Date,
+    val finishedAt: Date,
+    val applied: Int,
+    val failedToApply: Int,
+    val reconciled: Int,
+    val uploaded: Int,
+    val failedToUpload: Int,
+    val outgoingBatches: Int,
+    val failureReason: FailureReason?
+) {
+    companion object {
+        const val MILLIS_PER_SEC = 1000L
+
+        fun fromEngineInfo(uid: String, info: EngineInfo): BaseGleanSyncPing {
+            val failedToApply = info.incoming?.let {
+                it.failed + it.newFailed
+            } ?: 0
+            val (uploaded, failedToUpload) = info.outgoing.fold(Pair(0, 0)) { totals, batch ->
+                val (uploaded, failedToUpload) = totals
+                Pair(uploaded + batch.sent, failedToUpload + batch.failed)
+            }
+            return BaseGleanSyncPing(
+                uid = uid,
+                startedAt = Date(info.at.toLong() * MILLIS_PER_SEC),
+                // Glean intentionally doesn't support recording arbitrary
```

Comment:
nit: it might be useful for your team to have a reference to the issue here, https://bugzilla.mozilla.org/show_bug.cgi?id=1556341